csharp/ql/src/experimental/Security Features/JsonWebTokenHandler/JsonWebTokenHandlerLib.qll

```diff
@@ -0,0 +1,265 @@
+  le.getExpressionBody().(BoolLiteral).getBoolValue() = true
+}
+
+class CallableAlwaysReturnsTrue extends Callable {
+  CallableAlwaysReturnsTrue() {
+    callableHasAReturnStmtAndAlwaysReturnsTrue(this)
+    or
+    lambdaExprReturnsOnlyLiteralTrue(this)
+    or
+    exists(LambdaExpr le, Call call, CallableAlwaysReturnsTrue cat | this = le |
+      call = le.getExpressionBody() and
+      cat.getACall() = call
+    )
+  }
+}
+
+/**
+ * Holds if any exception being thrown by the callable is of type `System.ArgumentNullException`
+ * It will also hold if no exceptions are thrown by the callable
+ */
+predicate callableOnlyThrowsArgumentNullException(Callable c) {
+  forall(ThrowElement thre | c = thre.getEnclosingCallable() |
+    thre.getThrownExceptionType().hasQualifiedName("System.ArgumentNullException")
+  )
+}
+
+/**
+ * A specialization of `CallableAlwaysReturnsTrue` that takes into consideration exceptions being thrown for higher precision.
+ */
+class CallableAlwaysReturnsTrueHigherPrecision extends CallableAlwaysReturnsTrue {
+  CallableAlwaysReturnsTrueHigherPrecision() {
+    callableOnlyThrowsArgumentNullException(this) and
+    (
+      forall(Call call, Callable callable | call.getEnclosingCallable() = this |
+        callable.getACall() = call and
+        callable instanceof CallableAlwaysReturnsTrueHigherPrecision
+      )
+      or
+      exists(LambdaExpr le, Call call, CallableAlwaysReturnsTrueHigherPrecision cat | this = le |
+        call = le.getExpressionBody() and
+        cat.getACall() = call
+      )
+    )
+  }
+}
+
+/**
+ * A property Write for the `IssuerValidator` property for `Microsoft.IdentityModel.Tokens.TokenValidationParameters`
+ */
+class TokenValidationParametersPropertyWriteToValidationDelegatedIssuerValidator extends PropertyWrite {
+  TokenValidationParametersPropertyWriteToValidationDelegatedIssuerValidator() {
+    exists(Property p, Class c |
+      c.hasQualifiedName("Microsoft.IdentityModel.Tokens.TokenValidationParameters")
+    |
+      p.getAnAccess() = this and
+      c.getAProperty() = p and
+      p.hasName("IssuerValidator")
+    )
+  }
+}
+
+/**
+ * A callable that returns a `string` and has a `string` as 1st argument
+ */
+private class CallableReturnsStringAndArg0IsString extends Callable {
+  CallableReturnsStringAndArg0IsString() {
+    this.getReturnType() instanceof StringType and
+    this.getParameter(0).getType().toString() = "String"
+  }
+}
+
+/**
+ * A Callable that always return the 1st argument, both of `string` type
+ */
+class CallableAlwaysReturnsParameter0 extends CallableReturnsStringAndArg0IsString {
+  CallableAlwaysReturnsParameter0() {
+    forall(ReturnStmt rs | rs.getEnclosingCallable() = this |
+      rs.getChild(0) = this.getParameter(0).getAnAccess()
+    ) and
+    exists(ReturnStmt rs | rs.getEnclosingCallable() = this)
+    or
+    exists(LambdaExpr le, Call call, CallableAlwaysReturnsParameter0 cat | this = le |
```

Comment:
Expression-bodied `LambdaExpr` is special cased here. Should block bodied lambdas, and delegates be handled here as well (`AnonymousFunctionExpr`)?  (This applies to all the places where `LambdaExpr::getExpressionBody()` is used)

Reply:
Fixed